pyhf/__init__.py

```diff
@@ -308,52 +309,76 @@ def loglambdav(pars, data, pdf):
     return -2 * pdf.logpdf(pars, data)
 
 def qmu(mu, data, pdf, init_pars, par_bounds):
-    # The Test Statistic
-    mubhathat = tensorlib.tolist(optimizer.constrained_bestfit(loglambdav, mu, data, pdf, init_pars, par_bounds))
-    muhatbhat = tensorlib.tolist(optimizer.unconstrained_bestfit(loglambdav, data, pdf, init_pars, par_bounds))
-    qmu = tensorlib.tolist(loglambdav(mubhathat, data, pdf) - loglambdav(muhatbhat, data, pdf))[0]
-    if muhatbhat[pdf.config.poi_index] > mu:
-        return 0.0
+    """
+    The test statistic, q_mu, for establishing an upper
+    limit on the strength parameter, mu, as defiend in
+    Equation (14) in arXiv:1007.1727
+
+    .. math::
+       :nowrap:
+
+       q_{\mu} = \left\{\begin{array}{ll}
+       -2\ln\lambda\left(\mu\right), &\hat{\mu} < \mu,\\
+       0, & \hat{\mu} > \mu
+       \end{array}\right.
+
+    Args:
+        mu (Number or Tensor): The signal strength parameter
+        data (Tensor): The data to be considered
+        pdf (Tensor): The model used in the likelihood ratio calculation
+        init_pars (Tensor): The initial parameters
+        par_bounds(Tensor): The bounds on the paramter values
+
+    Returns:
+        Float: The calculated test statistic, q_mu
+    """
+    mubhathat = tensorlib.tolist(optimizer.constrained_bestfit(
+        loglambdav, mu, data, pdf, init_pars, par_bounds))
+    muhatbhat = tensorlib.tolist(optimizer.unconstrained_bestfit(
+        loglambdav, data, pdf, init_pars, par_bounds))
+    qmu = tensorlib.tolist(
+        loglambdav(mubhathat, data, pdf) - loglambdav(muhatbhat, data, pdf))[0]
+    qmu = tensorlib.tolist(
+        tensorlib.where(muhatbhat[pdf.config.poi_index] > mu, [0], qmu))
     return qmu
 
-from scipy.stats import norm
+
 def pvals_from_teststat(sqrtqmu_v, sqrtqmuA_v):
     # these pvals are from formula
     # (59) in arxiv:1007.1727 p_mu = 1-F(q_mu|mu') = 1- \Phi(q_mu - (mu-mu')/sigma)
     # and  (mu-mu')/sigma = sqrt(Lambda)= sqrt(q_mu_A)
-    CLsb = 1 - norm.cdf(sqrtqmu_v)
-    CLb =  1 - norm.cdf(sqrtqmu_v - sqrtqmuA_v)
+    CLsb = 1 - tensorlib.normal_cdf(sqrtqmu_v)
+    CLb = 1 - tensorlib.normal_cdf(sqrtqmu_v - sqrtqmuA_v)
     oneOverCLs = CLb / CLsb
     return CLsb, CLb, oneOverCLs
 
-import math
+
 def runOnePoint(muTest, data, pdf, init_pars, par_bounds):
-    asimov_mu = 0.0
-    asimov_data = tensorlib.tolist(generate_asimov_data(asimov_mu, data,
-                                   pdf, init_pars, par_bounds))
+    asimov_mu = 0.
+    asimov_data = tensorlib.tolist(
+        generate_asimov_data(asimov_mu, data, pdf, init_pars, par_bounds))
 
```

Comment:
probably also don't need `.tolist` here

Reply:
If we get rid of the `.tolist` here then we need to change other things. `asimov_data` gets passed to `qmu()` which then passes it to (for our supported optimizers) `optimizer.constrained_bestfit()`. This fails for PyTorch, as it requires a list for data. I would suggest removing the `.tolist` here and simply adding 

```python
data = self.tensorlib.tolist(data)
```
to `unconstrained_bestfit` and `constrained_bestfit` in `opt_pytorch`. Thoughts?

Reply:
thanks for the explanation. where exactly does that factor in? I only see it being passed to the objective function which is `pyhf.loglambdav` / `pdf.logpdf`  which I would have thought are fine with tensors

Reply:
Very good point. Indeed, it is fine with tensors (NumPy and TensorFlow handle it just fine). The problem arises in that for the PyTorch backend, in `generate_asimov_data()`, `pdf.expected_data()` is returning type `torch.autograd.variable.Variable` instead of `torch.FloatTensor`. 

If I stick a cast to `torch.FloatTensor` 

```python
data = torch.FloatTensor(self.tensorlib.tolist(data))
```

in just inside `constrained_bestfit` and `unconstrained_bestfit` this goes away, so this indicates that for the PyTorch backend `astensor()` is not giving the desired result.

```python
    def astensor(self, tensor_in):
        if isinstance(tensor_in, torch.autograd.Variable):
            v = tensor_in
        else:
            try:
                v = torch.autograd.Variable(torch.Tensor(tensor_in))
            except TypeError:
                try:
                    v = torch.autograd.Variable(tensor_in)
                except TypeError:
                    v = torch.autograd.Variable(torch.Tensor([tensor_in]))
                except RuntimeError:
                    v = torch.autograd.Variable(torch.Tensor([tensor_in]))
        return v.type(torch.FloatTensor)
```

Reply:
To make this a bit more easy to follow, I've pushed 305e561 (knowing this will fail the CI) so that the code that I'm talking about is at least in the PR.

Reply:
hi @matthewfeickert I tested this quick on `master` and see this. So I wonder how this breaks. Does that snippet work in your branch?

<img width="958" alt="screenshot" src="https://user-images.githubusercontent.com/2318083/39338837-a641ba78-49c5-11e8-8efc-289833a4c715.png">

Reply:
It does.
```python
>>> import pyhf
>>> import pyhf.simplemodels
>>> pyhf.set_backend(pyhf.tensor.pytorch_backend())
>>> pdf = pyhf.simplemodels.hepdata_like([10.], [50.], [7.])
>>> data = [50.] + pdf.config.auxdata
>>> asimov = pyhf.generate_asimov_data(0., data, pdf, pdf.config.suggested_init(), pdf.config.suggested_bounds())
>>> pdf.logpdf(pdf.config.suggested_init(), asimov)
Variable containing:
-6.7963
[torch.FloatTensor of size 1]
```

Reply:
just for posterity's sake: solution was to add `retain_graph` in backward() calls of the fit